pkg/service/schema.go

```diff
@@ -26,47 +30,152 @@ type bindingSchema struct {
 }
 
 type parametersSchema struct {
-	Schema     string                      `json:"$schema"`
-	Type       string                      `json:"type"`
-	Properties map[string]*ParameterSchema `json:"properties,omitempty"`
-	Required   []string                    `json:"required,omitempty"`
+	Schema             string                     `json:"$schema"`
+	Type               string                     `json:"type"`
+	Description        string                     `json:"description,omitempty"`
+	Required           bool                       `json:"-"`
+	RequiredProperties []string                   `json:"required,omitempty"`
+	Properties         map[string]ParameterSchema `json:"properties,omitempty"`
+	Additional         ParameterSchema            `json:"additionalProperties,omitempty"` // nolint: lll
 }
 
-// ParameterSchema represents the individual schema for a given parameter
-type ParameterSchema struct {
-	Type        string                      `json:"type"`
-	Required    bool                        `json:"-"`
-	Description string                      `json:"description,omitempty"`
-	Default     interface{}                 `json:"default,omitempty"`
-	Items       *ParameterSchema            `json:"items,omitempty"`
-	Properties  map[string]*ParameterSchema `json:"properties,omitempty"`
-	Additional  *ParameterSchema            `json:"additionalProperties,omitempty"`
+// SimpleParameterSchema represents the attributes of a simple schema type
+// such as a string or integer
+type SimpleParameterSchema struct {
+	Type          string      `json:"type"`
+	Description   string      `json:"description,omitempty"`
+	Required      bool        `json:"-"`
+	Default       interface{} `json:"default,omitempty"`
+	AllowedValues interface{} `json:"enum,omitempty"`
 }
 
-func (s *parametersSchema) addParameters(
-	newParams map[string]*ParameterSchema,
-) {
-	if s.Properties == nil {
-		s.Properties = newParams
-		for key, param := range newParams {
-			if param.Required {
-				s.Required = append(s.Required, key)
+// ObjectParameterSchema represents the attributes of a complicated schema type
+// that can have nested properties
+type ObjectParameterSchema struct {
+	Type               string                     `json:"type"`
```

Comment:
I get why `Type` is a field on `SimpleParameterSchema`, because in that case, the field could be a string, an integer, float, etc., but on `ObjectParameterSchema` and `ArrayParameterSchema`, shouldn't `Type` be implied by... the type?

I know it's required by the JSON, but maybe there's a way to hide it here? i.e. Hide it from users who are using these types to craft a schema?

I'm imagining that you could do something like remove `Type` here and then introduce a new, non-exported type `objectParameterSchema` that is composed of the exported `ObjectParameterSchema` and adds in the `Type` field. The same goes for `RequiredProperties`, which I believe you might like to also hide from module implementors.